src/datamodel/serializer/UUri.cpp

```diff
@@ -8,5 +8,193 @@
 // https://www.apache.org/licenses/LICENSE-2.0
 //
 // SPDX-License-Identifier: Apache-2.0
-
 #include "up-cpp/datamodel/serializer/UUri.h"
+
+#include <sstream>
+#include <vector>
+
+#include "up-cpp/datamodel/validator/UUri.h"
+
+namespace uprotocol::datamodel::serializer::uri {
+std::string AsString::serialize(const v1::UUri& uri) {
+	using namespace uprotocol::datamodel::validator::uri;
+	auto [valid, reason] = isValid(uri);
+	if (!valid) {
+		throw std::invalid_argument("Invalid UUri For Serialization | " +
+		                            std::string(message(*reason)));
+	}
+	std::stringstream ss;
+	ss << std::hex << std::uppercase;
+
+	if (!isLocal(uri)) {
+		ss << "//" << uri.authority_name();
+	}
+	ss << "/" << uri.ue_id() << "/" << uri.ue_version_major() << "/"
+	   << uri.resource_id();
+
+	return std::move(ss).str();
+}
+
+std::string_view extractSegment(std::string_view& uriView) {
+	const auto end = uriView.find("/");
+	if (end == uriView.npos) {
+		throw std::invalid_argument("Could not extract segment from '" +
+		                            std::string(uriView) +
+		                            "' with separator '" + "/" + "'");
+	}
+	auto segment = uriView.substr(0, end);
+	uriView = uriView.substr(end + 1);
+	return segment;
+}
+
+uint32_t segmentToUint32(const std::string_view& segment) {
+	uint32_t value = 0U;
+	std::istringstream iss{std::string(segment)};
```

Comment:
Allocates a copy of the string. This is why `from_chars` was used in the example I provided: it is specified to not produce copies.

Reply:
Done